### PR TITLE
gh-119127: functools.partial placeholders

### PR DESCRIPTION
**Pull request by dg-pb**
As I already had implementation I though PR might be helpful for others to see and evaluate.

From all the different extensions of `functools.partial` I think this one is the best. It is relatively simple and exposes all missing functionality. Other `partial` extensions that I have seen lack functionality and would not provide complete argument ordering capabilities and/or are too complicated in relation to what they offer.

Implementation can be summarised as follows:
a) All trailing placeholders are removed. (Makes things simpler)
b) Throws exception if not all placeholders are filled on call
c) retains optimization benefits of application on other `partial` instances.

Performance penalty compared to current `functools.partial` is minimal for extension class. + 20-30 ns for initialisation and <4 ns when called with or without placeholders.

To put it simply, new functionality extends `functools.partial` so that it has flexibility of `lambda` / `def` approach (in terms of argument ordering), but call overhead is 2x smaller.

The way I see it is that this could only be justified if this extension provided completeness and no new functionality is going to be needed anywhere near in the future. I have thought about it and tried various alternatives and I think there is a good chance that this is the case. Personally, I don't think I would ever need anything more from `partial` class.

Current implementation functions reliably.

* Issue: gh-119127

### CODE REVIEW

File: Modules/_functoolsmodule.c [hunk 1]
```diff
@@ -41,13 +41,31 @@ get_functools_state(PyObject *module)
...
+
+static PyTypeObject placeholder_type = {
+    .ob_base = PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "functools.Placeholder",
+    .tp_doc = PyDoc_STR("placeholder for partial arguments"),
+    .tp_basicsize = sizeof(placeholderobject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_DISALLOW_INSTANTIATION,
+    .tp_new = PyType_GenericNew,
+};
```
Comment (nineteendo) [inline Modules/_functoolsmodule.c]: https://github.com/python/cpython/actions/runs/9174174702/job/25224470051?pr=119303#step:16:69

> Non-constant global variables are generally not supported
in the CPython repo.

**Comment by dgrigonis** [comment]
I was thinking to wait for approval first.

**Comment by nineteendo** [comment]
You wrote "GN" in the title instead of "GH".

File: Modules/_functoolsmodule.c (hunk 1 above)
Comment (dg-pb) [reply Modules/_functoolsmodule.c]: I think it is because it didn't have `Py_TPFLAGS_IMMUTABLETYPE` flag, but can not check because the tool is not working for me localy.

Comment (nineteendo) [reply Modules/_functoolsmodule.c]: I think you can mark this as resolved.

Comment (dg-pb) [reply Modules/_functoolsmodule.c]: Finally! `c-analyzer` is not working for `clang` so could not run it locally.

**Comment by dgrigonis** [comment]
If this was accepted, the implementation is ready for review. Python and C versions are in sync and tests implemented.

File: Lib/test/test_typing.py
Comment (AlexWaygood) [inline Lib/test/test_typing.py:1]: Please revert changes that are unrelated to this PR's purpose

File: Lib/_pyrepl/simple_interact.py [hunk 2]
```diff
@@ -97,8 +97,8 @@ def runsource(self, source, filename="<input>", symbol="single"):
             try:
                 code = compile(item, filename, the_symbol)
             except (OverflowError, ValueError):
-                    self.showsyntaxerror(filename)
-                    return False
+                self.showsyntaxerror(filename)
+                return False
```
Comment (nineteendo) [inline Lib/_pyrepl/simple_interact.py]:
```suggestion
                    self.showsyntaxerror(filename)
                    return False
```

File: Lib/test/test_typing.py [hunk 3]
```diff
@@ -7345,7 +7345,7 @@ def test_no_generator_instantiation(self):
 
     def test_async_generator(self):
         async def f():
-             yield 42
+            yield 42
```
Comment (nineteendo) [inline Lib/test/test_typing.py]:
```suggestion
             yield 42
```

**Review by nineteendo — COMMENTED** [review]
Revert unrelated changes, as demanded by @AlexWaygood.

**Comment by nineteendo** [comment]
@dgrigonis could you re-do your pull request? It has become impossible to see what you changed.

**Comment by erlend-aasland** [comment]
Please open a new clean PR against main. Make sure the diff against main is correct before opening the PR.

**Comment by nineteendo** [comment]
Thanks Erlend.
